GitHub app blog (#2988)
jenkins-infra/jenkins.io: content/blog/2020/04/2020-04-16-github-app-authentication.adoc

---
layout: post

> title: "GitHub app authentication released"

Reviewer: ```suggestion
title: "GitHub App authentication support released"
```

Reviewer: Should we also scope it to multi-branch Pipeline?

tags:
- github

> - github-branch-source

Reviewer: ```suggestion
- github-branch-source
- pipeline
- announcement
```

> author: timja

Reviewer: Some social media image would be great. E.g we could simply emulate the GitHub Web UI:

![image](https://user-images.githubusercontent.com/3000480/79386750-e3015e00-7f6a-11ea-8fea-6e15b5b3a8f6.png)

Source: https://docs.google.com/presentation/d/1ncGvxS3KX1JesCg9FJ04y94jdEk9mLKWbbVwBUbyPuo/edit?usp=sharing

---

> I'm excited to announce support for authenticating as a GitHub app in Jenkins.

Reviewer: ```suggestion
I'm excited to announce support for authenticating as a GitHub app in Jenkins.
This has been a long awaited feature by many users.
```

> It has been released in link:https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.7.0-beta1[GitHub Branch Source 2.7.0-beta1]. This has been a long awaited feature by many users.

Reviewer: ```suggestion
It has been released in link:https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.7.0-beta1[GitHub Branch Source 2.7.0-beta1] which is available in the link:/doc/developer/publishing/releasing-experimental-updates/#using-the-experimental-update-center[Jenkins experimental update center].
```

 Authenticating as a GitHub app brings many benefits:

* **Larger rate limits** - The rate limit for a GitHub app scales with your organization size, 
whereas a user based token has a limit of 5000 regardless of how many repositories you have.

* ** User-independent authentication** - Each GitHub app has its own user-independent authentication. No more need for 'bot' users or figuring out who should be the owner of 2FA or OAuth tokens.

> * **Improved security and tighter permissions** - GitHub Apps offers much finer-grained permissions compared to a service user and its personal access tokens. This lets the Jenkins GitHub app require a much smaller set of privileges to run properly.

Reviewer: ```suggestion
* **Improved security and tighter permissions** - GitHub Apps offer much finer-grained permissions compared to a service user and its personal access tokens. This lets the Jenkins GitHub app require a much smaller set of privileges to run properly.
```

* **Access to GitHub Checks API** - GitHub Apps can access the the  link:https://developer.github.com/v3/checks/[GitHub Checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation

== Getting started

Install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin],

> make sure the version is at least 2.7.0-beta1. 

Reviewer: ```suggestion
make sure the version is at least `2.7.0-beta1`. 
Installation guidelines for beta releases are available link:/doc/developer/publishing/releasing-experimental-updates/#using-the-experimental-update-center[here]
```

I will document Docker there

> Then follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[setup guide].  These instructions are also linked from the plugin’s README on GitHub.

Reviewer: ```suggestion
Then follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[GitHub App Authentication setup guide].  These instructions are also linked from the plugin’s README on GitHub.
```

Once you've finished setting it up, Jenkins will validate your credential and you should see your new rate limit.
Here's an example on a large org:

image:/images/github-app-rate-limit.png[GitHub app rate limit]

> 

Reviewer: Would it make sense to emphasize that GitHub Branch Source will use App authentication once it is completed?

Author: what do you mean?

Reviewer: I mean using App rate limit when doing the repository scans

Reviewer: Maybe this is enough: https://github.com/jenkins-infra/jenkins.io/pull/2988/files#r409133696

== How do I get an API token in my pipeline?

> You can access the Bearer token for the GitHub API by just loading a 'Username/Password' credential as usual,

Reviewer: ```suggestion
In addition to usage of GitHub App authentication for Multi-Branch Pipeline, you can also app authentication directly in your Pipelines. 
You can access the Bearer token for the GitHub API by just loading a 'Username/Password' credential as usual,
```

the plugin will handle authenticating with GitHub in the background.

This could be used to call additional GitHub API endpoints from your pipeline, possibly the
link:https://developer.github.com/v3/repos/deployments/[deployments api] or you may wish to implement your own
links:https://developer.github.com/v3/checks/[checks api] integration until Jenkins supports this out of the box.

> _Note: the api token you get will only be valid for one hour, don't get it at the start of the pipeline and assume it will be valid all the way through_

Reviewer: ```suggestion
_Note: the API token you get will only be valid for one hour, don't get it at the start of the pipeline and assume it will be valid all the way through_
```

> 

Reviewer: ```suggestion

**Example:** Let's submit a check run to Jenkins from our Pipeline:

```

[source, groovy]
----

pipeline {
  agent any

  stages{
    stage('Check run') { 
      steps {
        withCredentials([usernamePassword(credentialsId: 'githubapp-jenkins',
                                          usernameVariable: 'GITHUB_APP',
                                          passwordVariable: 'GITHUB_JWT_TOKEN')]) {
            sh '''
            curl -H "Content-Type: application/json" \
                 -H "Accept: application/vnd.github.antiope-preview+json" \
                 -H "authorization: Bearer ${GITHUB_JWT_TOKEN}" \
                 -d '{ "name": "check_run", \
                       "head_sha": "'${GIT_COMMIT}'", \
                       "status": "in_progress", \
                       "external_id": "42", \
                       "started_at": "2020-03-05T11:14:52Z", \
                       "output": { "title": "Check run from Jenkins!", \
                                   "summary": "This is a check run which has been generated from Jenkins as GitHub App", \
                                   "text": "...and that is awesome"}}' https://api.github.com/repos/<org>/<repo>/check-runs
            '''
        }
      }
    }
  }
}

----

== What's next

> GitHub Apps authentication in Jenkins is a huge improvement.  Many teams have already started using it and have helped improve it by giving pre-release feedback. There are more improvements on the way. 

Reviewer: ```suggestion
GitHub Apps authentication in Jenkins is a huge improvement.  Many teams have already started using it and have helped improve it by giving pre-release feedback. There are more improvements on the way.
I hope that the feature will be released shortly in the main Jenkins update center once they are released.
```

There's a proposed Google Summer of Code project: link:https://jenkins.io/projects/gsoc/2020/project-ideas/github-checks/[GitHub Checks API for Jenkins Plugins].  

> It will look at integrating with the link:https://developer.github.com/v3/checks/[checks api], 

Reviewer: ```suggestion
It will look at integrating with the link:https://developer.github.com/v3/checks/[Checks API], 
```

with a focus on reporting issues found using the link:https://plugins.jenkins.io/warnings-ng/[warnings-ng plugin]
directly onto the GitHub pull requests, along with test results summary on GitHub.

> If you want to get involved with this, join the link:https://gitter.im/jenkinsci/gsoc-sig[gitter channel]

Reviewer: ```suggestion
Hopefully it will make the Pipeline example below much simpler for Jenkins users :) 
If you want to get involved with this, join the link:https://gitter.im/jenkinsci/gsoc-sig[GSoC Gitter channel]
```

whereas a user based token has a limit of 5000 regardless of how many repositories you have.

* ** User-independent authentication** - Each GitHub app has its own user-independent authentication. No more need for 'bot' users or figuring out who should be the owner of 2FA or OAuth tokens.

* **Improved security and tighter permissions** - GitHub Apps offer much finer-grained permissions compared to a service user and its personal access tokens. This lets the Jenkins GitHub app require a much smaller set of privileges to run properly.

* **Access to GitHub Checks API** - GitHub Apps can access the the  link:https://developer.github.com/v3/checks/[GitHub Checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation

== Getting started

Install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin],
make sure the version is at least `2.7.0-beta1`. 
Installation guidelines for beta releases are available link:/doc/developer/publishing/releasing-experimental-updates/#using-the-experimental-update-center[here]

> Then follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[GitHub App Authentication setup guide].  These instructions are also linked from the plugin’s README on GitHub.

Reviewer: ```suggestion

== Configuring the github organization folder

Then follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[GitHub App Authentication setup guide].  These instructions are also linked from the plugin’s README on GitHub.
```
